File: app/src/main/java/com/hencoder/hencoderpracticedraw1/practice/Practice2DrawCircleView.java
package com.hencoder.hencoderpracticedraw1.practice;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.View;

import com.hencoder.hencoderpracticedraw1.Utils;

public class Practice2DrawCircleView extends View {

    Paint mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    public Practice2DrawCircleView(Context context) {
        super(context);
    }

    public Practice2DrawCircleView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public Practice2DrawCircleView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

//        练习内容：使用 canvas.drawCircle() 方法画圆
//        一共四个圆：1.实心圆 2.空心圆 3.蓝色实心圆 4.线宽为 20 的空心圆
        //1
        canvas.drawCircle(300,300,100,mPaint);
        //2
        mPaint.setStyle(Style.STROKE);
        canvas.drawCircle(500,300,100,mPaint);
        //3
        mPaint.setStyle(Style.FILL);
        mPaint.setColor(Color.BLUE);
        canvas.drawCircle(300, 500, 100, mPaint);
        //4
        mPaint.setStrokeWidth(Utils.dp2px(20));
        mPaint.setStyle(Style.STROKE);
        mPaint.setColor(Color.BLACK);
        canvas.drawCircle(600, 600, 100, mPaint);
    }
}

File: app/src/main/java/com/hencoder/hencoderpracticedraw1/practice/Practice2DrawCircleView.kt
package com.hencoder.hencoderpracticedraw1.practice

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View

/**
 * Created by xh on 2017/9/15.
 */
class Practice2DrawCircleView : View {

    private var mPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    private val mRadius = 150f

    private val mMargins = 70f

    private var mLeftMargin = 0f

    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    init {
        mPaint.strokeWidth = 3f
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        super.onSizeChanged(w, h, oldW, oldH)
        mLeftMargin = (w - 2 * mRadius - mMargins) / 2
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        mPaint.color = Color.BLACK
        canvas.drawCircle(mLeftMargin, mRadius, mRadius, mPaint)

        mPaint.style = Paint.Style.STROKE
        canvas.drawCircle(mLeftMargin + 2 * mRadius + mMargins, mRadius, mRadius, mPaint)


        mPaint.color = Color.parseColor("#FF4A90E2")
        mPaint.style = Paint.Style.FILL
        canvas.drawCircle(mLeftMargin, mRadius * 3 + mMargins, mRadius, mPaint)

        mPaint.color = Color.BLACK
        mPaint.style = Paint.Style.STROKE
        mPaint.strokeWidth = 60F
        canvas.drawCircle(mLeftMargin + 2 * mRadius + mMargins, mRadius * 3 + mMargins, mRadius, mPaint)

    }

}